File: leetcode/POTD/2025-02-07/solution.ts
function queryResults(limit: number, queries: number[][]): number[] {
	const result: number[] = []
	const ballToColor = new Map<number, number>()
	const colorCount = new Map<number, number>()

	for (const [ballId, newColor] of queries) {
		if (ballToColor.has(ballId)) {
			const previousColor = ballToColor.get(ballId)!
			colorCount.set(previousColor, colorCount.get(previousColor)! - 1)
			if (colorCount.get(previousColor) === 0) {
				colorCount.delete(previousColor)
			}
		}

		ballToColor.set(ballId, newColor)
		colorCount.set(newColor, (colorCount.get(newColor) || 0) + 1)
		result.push(colorCount.size)
	}

	return result
}

File: leetcode/POTD/2025-02-07/solution.java
class Solution {
    public int[] queryResults(int limit, int[][] queries) {
        int[] result = new int[queries.length];
        Map<Integer, Integer> ballToColor = new HashMap<>();
        Map<Integer, Integer> colorCount = new HashMap<>();

        for (int i = 0; i < queries.length; i++) {
            int ballId = queries[i][0];
            int newColor = queries[i][1];

            if (ballToColor.containsKey(ballId)) {
                int previousColor = ballToColor.get(ballId);
                colorCount.put(previousColor, colorCount.get(previousColor) - 1);
                if (colorCount.get(previousColor) == 0) {
                    colorCount.remove(previousColor);
                }
            }

            ballToColor.put(ballId, newColor);
            colorCount.put(newColor, colorCount.getOrDefault(newColor, 0) + 1);
            result[i] = colorCount.size();
        }

        return result;
    }
}